### PR TITLE
Feature/set value should not set `dirty` to true and give users options

### PR DESCRIPTION
**Pull request by bluebill1049**
`setValue` API refinement

**Reason**

The current API is confusing with support both name as string and array of values. In addition, programmatically set a field should give users the option to set `dirty`, we shouldn't set dirty to `true` when user setValue programmatically by default. 

`setValue(fieldName, values, config)`

---

`setValue('test', 'value',  { shouldValidate: false, shouldDirty: false })`

### CODE REVIEW

File: src/useForm.test.tsx [hunk 1]
```diff
@@ -1155,24 +1108,29 @@ describe('useForm', () => {
...
-              'test[0].name': 'default_update',
-              'test[1].name': 'default1',
-              'test[2].name': 'default2',
-            },
-          ]);
-          expect(result.current.formState.dirtyFields.test!).toEqual([
-            { name: true },
-          ]);
+          expect(transformToNestObject).toBeCalled();
+          // expect((transformToNestObject as any).mock.calls[2]).toEqual([
```
Comment (JeromeDeLeon) [inline src/useForm.test.tsx:1124]: Have you forgot to delete this or you have something still in your mind? :wink:

**Review by JeromeDeLeon — DISMISSED** [review]
Do you have something in mind in case users wanted to set multiple values?

File: src/useForm.test.tsx (hunk 1 above)
Comment (bluebill1049) [reply src/useForm.test.tsx:1124]: lol! good pick up.

**Comment by bluebill1049** [comment]
> Do you have something in mind in case users wanted to set multiple values?

they have to invoke multiple times or use `reset()` with `getValues()`

File: src/useForm.ts [hunk 2]
```diff
@@ -411,6 +371,56 @@ export function useForm<
...
+        if (field) {
+          setFieldValue(field, value[key]);
+
+          if (config.shouldDirty) {
+            setDirty(fieldName);
+          }
+
+          if (config.shouldValidate) {
+            trigger(fieldName as FieldName<TFieldValues>);
+          }
```
Comment (keiya01) [inline src/useForm.ts:400]: I think this condition is unnecessary. Because `trigger` method support array field.

**Review by keiya01 — COMMENTED** [review]
I have only one question. Other than that, it looks good to me!

File: src/useForm.ts (hunk 2 above)
Comment (bluebill1049) [reply src/useForm.ts:400]: Thanks @keiya01 which condition? `if` ?

Comment (keiya01) [reply src/useForm.ts:400]: Yes, `if`.

**Comment by JeromeDeLeon** [comment]
I think its important to let users know of this although I agree with you their. I have never use `setValue` with multiple values because the word says it all, `setValue` not `setValues`, :smile:

**Comment by bluebill1049** [comment]
> I think its important to let users know of this although I agree with you their. I have never use `setValue` with multiple values because the word says it all, `setValue` not `setValues`, 😄

agree,  i have already updated the doc.

File: src/useForm.ts (hunk 2 above)
Comment (bluebill1049) [reply src/useForm.ts:400]: without it will validate input each time users set a value, it's actually optional

Comment (bluebill1049) [reply src/useForm.ts:400]: by the way, your unit tests r so good!!

Comment (bluebill1049) [reply src/useForm.ts:400]: check this out: https://coveralls.io/github/bluebill1049/react-hook-form?branch=master

Comment (keiya01) [reply src/useForm.ts:400]: Oh... I understood. Sorry.

**Review by keiya01 — APPROVED** [review]
LGTM :)

**Comment by jviall** [comment]
@bluebill1049 Hi there, I'm confused because I can't seem to get this functionality to work for me. Particularly the setting a field to `dirty === false` when using `setValue`. Here's a sandbox forked from one of the doc's examples where I try to use `setValue` to "reset" an individual field back to its default value:

https://codesandbox.io/s/react-hook-form-formstate-dirty-touched-submitted-forked-opflm?file=/src/index.js

**Comment by bluebill1049** [comment]
@jviall can you submit a GitHub discussion question instead?

**Pull request merged**